resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/kerberos/KerberosPVWatcherCache.scala

```diff
@@ -0,0 +1,184 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.k8s.integrationtest.kerberos
+
+import java.util.concurrent.locks.{Condition, Lock, ReentrantLock}
+
+import scala.collection.JavaConverters._
+
+import io.fabric8.kubernetes.api.model.{PersistentVolume, PersistentVolumeClaim}
+import io.fabric8.kubernetes.client.{KubernetesClientException, Watch, Watcher}
+import io.fabric8.kubernetes.client.Watcher.Action
+
+import org.apache.spark.internal.Logging
+
+ /**
+  * This class is responsible for ensuring that the persistent volume claims are bounded
+  * to the correct persistent volume and that they are both created before launching the
+  * pods which expect to use them.
+  */
+private[spark] class KerberosPVWatcherCache(
+    kerberosUtils: KerberosUtils,
+    labels: Map[String, String]) extends Logging {
+    private val kubernetesClient = kerberosUtils.getClient
+    private val namespace = kerberosUtils.getNamespace
+    private var pvWatcher: Watch = _
+    private var pvcWatcher: Watch = _
+    private var pvCache =
+      scala.collection.mutable.Map[String, String]()
+    private var pvcCache =
+      scala.collection.mutable.Map[String, String]()
+    private var lock: Lock = new ReentrantLock()
+    private var nnBounded: Condition = lock.newCondition()
+    private var ktBounded: Condition = lock.newCondition()
```

Comment:
I would be more comfortable with using `Future`s instead of Condition locks, but perhaps that's a matter of preference - I'm not sure if there are any precedents in the codebase we can follow.